### ISSUE TITLE
MNT Bump requirements for compatibility with python 3.9

### ISSUE THREAD

**Issue by romanlutz** [fairlearn/fairlearn#915]
Some of our min requirements don't exist in Python 3.9, e.g., scikit-learn seems to have adopted python 3.9 in version 0.24, whereas our nightly "requirements-fixed" build runs with our minimum requirement of 0.22.1 and fails accordingly.

Should we bump our min requirements to

```
numpy>=1.20.0
pandas>=1.2.0
scikit-learn>=0.24.0
scipy>=1.6.0
```
from currently
```
numpy>=1.17.2
pandas>=0.25.1
scikit-learn>=0.22.1
scipy>=1.4.1
```
(and probably several others in requirements-dev.txt as well as `matplotlib>=3.4.0` for requirements-customplots.txt)?

I'm generally in favor of moving forward rather than staying on old versions, but I know there are people who are forced into old versions of some packages that might not be able to use our latest version as a consequence.

@fairlearn/fairlearn-maintainers Wdyt?

**Comment by adrinjalali** [comment]
Min requirements are usually not the same for all versions of python. To me, min requirements are tested with the oldest python we support, not with the newest one. Although I'm usually on the side of having newer dependencies, but I don't think we need to update what `pip install fairlearn` _requires_ just because the situation under the latest python says so.

**Comment by hildeweerts** [comment]
I think I'd prefer a situation where having the minimum requirements installed will always work, which is not the case for the current requirements, right? I'm not sure whether that's realistic though... How does scikit-learn deal with this, @adrinjalali?

**Comment by romanlutz** [comment]
@hildeweerts that's what I assumed as the expected situation before @adrinjalali's post. I do see the benefit in what @adrinjalali suggested, since it's not actually guaranteed that the latest working version of a package supports both Python 3.6 AND Python 3.9 (if they've already moved on to 3.7). We could have a python version-specific set of requirements, but that's not ideal either and doesn't gel well with PyPI which only takes one package upload (rather than n for n different Python versions).

So in conclusion I am in support of @adrinjalali's proposal of pinning the requirements-fixed build to the oldest python version we support (currently 3.6), even if the package versions used there may not even exist in the latest version we support (currently 3.9).

**Comment by hildeweerts** [comment]
Thank you for explaining it further @romanlutz. Sounds very reasonable. 

A related question: how do we decide to drop support for a particular version of Python?

**Comment by romanlutz** [comment]
We don't 🙂 We just follow what NumPy does: https://numpy.org/neps/nep-0029-deprecation_policy.html
In particular:
```
On Dec 26, 2021 drop support for Python 3.7 (initially released on Jun 27, 2018)
...
On Apr 14, 2023 drop support for Python 3.8 (initially released on Oct 14, 2019)
```

**Comment by romanlutz** [comment]
In conclusion, this would mean that the required change is to change the following line in `nightly-requirements-fixed.yml` from
```
    pyVersions: [3.6, 3.7, 3.8, 3.9]
```
to
```
    pyVersions: [3.6, 3.7, 3.8]
```

I'll tag this as an issue for new contributors, but would also love to hear if @riedgar-ms has any objections.

**Comment by adrinjalali** [comment]
Für each Python version we should have the oldest dependencies we support, I would object dropping Python 3.9 here. Alternatively, we can do one "min dependency" check and only include Python 3.6

**Comment by romanlutz** [comment]
I'm confused. I don't see scikit-learn testing this with per-version dependencies. https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/_min_dependencies.py

What you wrote sounds to me like we need multiple sets of dependency versions, because there's no way (that I'm aware of) to `pip install --min-version` or something like that. That's why @riedgar-ms 's script reads the version from the file and ensures we use the oldest one possible using `==` rather than `>=`.

**Comment by adrinjalali** [comment]
@romanlutz I mean if we're going to test the min dependencies per python version. AFAIK in sklearn we only check for the oldest deps on the oldest python we support (i.e. 3.7): https://github.com/scikit-learn/scikit-learn/blob/3c732b9f6a77e95dfa6beb154ca2e1e7848b74f9/azure-pipelines.yml#L167-L174

**Comment by riedgar-ms** [comment]
Sorry, should have commented before. I agree that maintaining separate minimum requirements per Python version is going to be a bit of a pain.

@adrinjalali  am I right in thinking that your suggestion is essentially to reduce the `nightly-fixed` set to `pyVersions: [3.6]` ? I can live with that.

**Comment by adrinjalali** [comment]
I'd vote for 3.7, but yes, having only one python version there is what I mean.

**Comment by romanlutz** [comment]
Sounds good. So what I wrote [earlier](https://github.com/fairlearn/fairlearn/issues/915#issuecomment-894075107) wasn't all that far off, then!

**Comment by hildeweerts** [comment]
> Sounds good. So what I wrote [earlier](https://github.com/fairlearn/fairlearn/issues/915#issuecomment-894075107) wasn't all that far off, then!

(Perhaps you can update your reply to what is expected now for the potential new contributor who wants to pick this up?)

**Comment by sushmitaS16** [comment]
Hello, I am a newbie to opensource. I followed your conversation above and am interested to contribute. Could you please tell me where exactly to start?  
Thanks

**Comment by romanlutz** [comment]
@sushmitaS16 You can follow the steps outlined in [this post](https://github.com/fairlearn/fairlearn/issues/915#issuecomment-894075107). 

@hildeweerts The described lines there are fine. It does pass for all three listed versions, just 3.9 is failing. In the future we'll eventually remove 3.6 support at which point we can remove it from here, too. But there's no particular benefit in restricting to just 3.6 or 3.7. We do know it works for all three.

**Comment by sushmitaS16** [comment]
Thanks @romanlutz 
Is the file mentioned [here](https://github.com/fairlearn/fairlearn/issues/915#issuecomment-894075107) the only one which needs to be edited, or do I need to modify any other .yml file?

**Comment by romanlutz** [comment]
It's the only one 🙂 I will kick off a run of that pipeline once you have a pull request, just to be sure it works fine. Thanks again for picking this up!
